controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,887 @@
+  // These constants are called N and M in the reference manual:
+  //
+  // Fin = 4MHz
+  // Fvco = Fin * (N/M)
+  //
+  // Legal range for Fvco is 96MHz to 344MHz according to the
+  // data sheet.  I'll use 160MHz for Fvco and divide by 2
+  // to get an 80MHz output clock
+  //
+  // See chapter 6 of the reference manual
+  int N = 40;
+  int M = 1;
+  RCC_Regs *rcc = RCC_BASE;
+  rcc->pllCfg = 0x01000001 | (N << 8) | ((M - 1) << 4);
+
+  // Turn on the PLL
+  rcc->clkCtrl |= 0x01000000;
+
+  // Wait for the PLL ready indication
+  while (!(rcc->clkCtrl & 0x02000000)) {
+  }
+
+  // Set PLL as system clock
+  rcc->clkCfg = 0x00000003;
+
+  // Use system clock as the A/D clock
+  rcc->indClkCfg = 0x30000000;
+}
+
+/*
+ * One time init of HAL.
+ */
+void HalApi::init() {
+
+  // Init various components needed by the system.
+  InitGPIO();
+  InitSysTimer();
+  InitADC();
+  InitPwmOut();
+  InitUARTs();
+  watchdog_init();
+
+  // Enable interrupts
+  IntEnable();
+}
+
+// Reset the processor
+[[noreturn]] void HalApi::reset_device() {
+
+  // Note that the system control registers are a standard ARM peripherial
+  // they aren't documented in the normal STM32 reference manual, rather
+  // they're in the processor programming manual.
+  // The register we use to reset the system is called the
+  // "Application interrupt and reset control register (AIRCR)"
+  SysCtrl_Reg *sysCtl = SYSCTL_BASE;
+  sysCtl->apInt = 0x05FA0004;
+
+  // We promised we wouldn't return, so...
+  while (true) {
+  }
+}
+
+/******************************************************************
+ * General Purpose I/O support.
+ *
+ * The following pins are used as GPIO on the rev-1 PCB
+ *
+ * ID inputs.  These can be used to identify the PCB revision
+ * we're running on.
+ *  PB1  - ID0
+ *  PA12 - ID1
+ *
+ * LED outputs.
+ *  PC13 - red
+ *  PC14 - yellow
+ *  PC15 - green
+ *
+ * Solenoid
+ *  PA11 - Note, this is also a timer pin so we may want to
+ *         PWM it to reduce the solenoid voltage.
+ *         For no I'm treating it as a digital output.
+ *****************************************************************/
+static void InitGPIO() {
+  // See chapter 8 of the reference manual for details on GPIO
+
+  // Enable all the GPIO clocks
+  EnableClock(GPIO_A_BASE);
+  EnableClock(GPIO_B_BASE);
+  EnableClock(GPIO_C_BASE);
+  EnableClock(GPIO_D_BASE);
+  EnableClock(GPIO_E_BASE);
+  EnableClock(GPIO_H_BASE);
+
+  // Configure PCB ID pins as inputs.
+  GPIO_PinMode(GPIO_B_BASE, 1, GPIO_MODE_INPUT);
+  GPIO_PinMode(GPIO_A_BASE, 12, GPIO_MODE_INPUT);
+
+  // Configure LED pins as outputs
+  GPIO_PinMode(GPIO_C_BASE, 13, GPIO_MODE_OUTPUT);
+  GPIO_PinMode(GPIO_C_BASE, 14, GPIO_MODE_OUTPUT);
+  GPIO_PinMode(GPIO_C_BASE, 15, GPIO_MODE_OUTPUT);
+
+  // Turn all three LEDs off initially
+  GPIO_ClrPin(GPIO_C_BASE, 13);
+  GPIO_ClrPin(GPIO_C_BASE, 14);
+  GPIO_ClrPin(GPIO_C_BASE, 15);
+
+  // Configure the solenoid and turn it off
```

Comment:
These pin numbers are only used a couple times in the HAL (once when initialized, once when changed) and they're right next to each other.  I don't know if adding additional constants would improve the clarity.  Also, it's a bit unwieldy because it's not just the pin number, it's also the bank (GPIO_A_BASE) which is no longer a constant, it's a pointer.